component/xload/comp/lister.go

```diff
@@ -0,0 +1,198 @@
+			path: path,
+		},
+		listBlocked: false,
+	}
+
+	rl.SetName(LISTER)
+	rl.SetRemote(remote)
+	rl.Init()
+	return rl, nil
+}
+
+func (rl *remoteLister) Init() {
+	rl.SetThreadPool(common.NewThreadPool(common.MAX_LISTER, rl.Process))
+	if rl.GetThreadPool() == nil {
+		log.Err("remoteLister::Init : fail to init thread pool")
+	}
+}
+
+func (rl *remoteLister) Start() {
+	log.Debug("remoteLister::Start : start remote lister for %s", rl.path)
+	rl.GetThreadPool().Start()
+	rl.GetThreadPool().Schedule(&common.WorkItem{CompName: rl.GetName()})
+}
+
+func (rl *remoteLister) Stop() {
+	log.Debug("remoteLister::Stop : stop remote lister for %s", rl.path)
+	if rl.GetThreadPool() != nil {
+		rl.GetThreadPool().Stop()
+	}
+	rl.GetNext().Stop()
+}
+
+// wait for the configured block-list-on-mount-sec to make the list call
+func waitForListTimeout() error {
+	var blockListSeconds uint16 = 0
+	err := config.UnmarshalKey("azstorage.block-list-on-mount-sec", &blockListSeconds)
+	if err != nil {
+		return err
+	}
+	time.Sleep(time.Duration(blockListSeconds) * time.Second)
+	return nil
+}
+
+func (rl *remoteLister) Process(item *common.WorkItem) (int, error) {
+	absPath := item.Path // TODO:: xload : check this for subdirectory mounting
+
+	log.Debug("remoteLister::Process : Reading remote dir %s", absPath)
+
+	// this block will be executed only in the first list call for the remote directory
+	// so haven't made the listBlocked variable atomic
+	if !rl.listBlocked {
+		log.Debug("remoteLister::Process : Waiting for block-list-on-mount-sec before making the list call")
+		err := waitForListTimeout()
+		if err != nil {
+			log.Err("remoteLister::Process : unable to unmarshal block-list-on-mount-sec [%s]", err.Error())
+			return 0, err
+		}
+		rl.listBlocked = true
+	}
+
+	marker := ""
+	var cnt, iteration int
+	for {
+		entries, new_marker, err := rl.GetRemote().StreamDir(internal.StreamDirOptions{
+			Name:  absPath,
+			Token: marker,
+		})
+		if err != nil {
+			log.Err("remoteLister::Process : Remote listing failed for %s [%s]", absPath, err.Error())
+		}
+
+		marker = new_marker
+		cnt += len(entries)
+		iteration++
+		log.Debug("remoteLister::Process : count: %d , iterations: %d", cnt, iteration)
+
+		for _, entry := range entries {
+			log.Debug("remoteLister::Process : Iterating: %s, Is directory: %v", entry.Path, entry.IsDir())
+
+			if entry.IsDir() {
+				// create directory in local
+				// spawn go routine for directory creation and then
+				// adding to the input channel of the listing component
+				// TODO:: xload : check how many threads can we spawn
+				go func(name string) {
+					localPath := filepath.Join(rl.path, name)
+					err = rl.mkdir(localPath)
+					// TODO:: xload : handle error
+					if err != nil {
+						log.Err("remoteLister::Process : Failed to create directory [%s]", err.Error())
+						return
+					}
+
+					// push the directory to input pool for its listing
+					rl.GetThreadPool().Schedule(&common.WorkItem{
```

Comment:
Component pushing back into self queue shall have a check on the channel size. If channel is full its going to hand up and then we may end up having many threads in this state. Better to have a log if the channel is already full so that if we hit any such issue on customer setup its easier to debug.